farm/modeling/predictions.py

```diff
@@ -1,77 +1,166 @@
+        # TODO THIS IS A LOT OF ATTRIBUTES - REMOVE SOME WITH METHODS
+        # self.answer_type can be "is_impossible", "yes", "no" or "span"
+        self.answer_type = answer_type
         self.score = score
-        self.unit = unit
+        self.probability = probability
+
+        # If self.answer_type is "span", self.answer is a string answer span
+        # Otherwise, it is None
+        self.answer = answer
+        self.offset_answer_start = offset_answer_start
+        self.offset_answer_end = offset_answer_end
+
+        # If self.answer_type is in ["yes", "no"] then self.answer_support is a text string
+        # If self.answer is a string answer span or self.answer_type is "is_impossible", answer_support is None
+        self.answer_support = answer_support
+        self.offset_answer_support_start = offset_answer_support_start
+        self.offset_answer_support_end = offset_answer_support_end
         self.sample_idx = sample_idx
-        self.classification = classification
-        self.n_samples = n_samples
-        self.pred_str = pred_str
-        self.id = id
-        self.level = level
 
-    def to_list(self):
-        return [self.pred_str, self.start, self.end, self.score, self.sample_idx]
+        # self.context is the document or passage where the answer is found
+        self.context = context
+        self.offset_context_start = offset_context_start
+        self.offset_context_end = offset_context_end
+
+        # Offset unit is either "token" or "char"
+        # Aggregation level is either "doc" or "passage"
+        self.offset_unit  = offset_unit
+        self.aggregation_level = aggregation_level
+
+        self.document_id = document_id
+        self.passage_id = passage_id
+        # TODO This seems to be necessary in order to align QA preds of len n_docs with Text Classification preds
+        # TODO of len n_passages
+        self.n_samples_in_doc = n_samples_in_doc
 
-    def __str__(self):
-        if self.pred_str is None:
-            pred_str = "is_impossible"
+    def to_doc_level(self, start, end):
+        self.offset_answer_start = start
+        self.offset_answer_end = end
+        self.aggregation_level = "document"
+
+    def add_answer(self, string):
+        if string == "":
+            self.answer = "is_impossible"
+            assert self.offset_answer_end == -1
+            assert self.offset_answer_start == -1
         else:
-            pred_str = self.pred_str
-        ret = f"answer: {pred_str}\n" \
-              f"score: {self.score}"
-        return ret
+            self.answer = string
+            assert self.offset_answer_end >= 0
+            assert self.offset_answer_start >= 0
 
-    def __repr__(self):
-        return str(self)
 
-class DocumentPred:
-    """ Contains a collection of Span predictions for one document. Used in Question Answering. Also contains all
-    attributes needed to generate the appropriate output json"""
+
+class QAPred(Pred):
```

Comment:
Can we add type annotations here, too please?

Reply:
Yes sorry forgot to do this object too